### PR TITLE
buildImageButton: add support of building image with mulitple targets <HMS-4066>

### PR DESCRIPTION
**Pull request by mgold1234**
this commit add support of building an image with mulitple targets, when user build an image, he can choose which targets he wants to build.

### CODE REVIEW

File: src/Components/CreateImageWizardV2/steps/Review/Footer/CreateDropdown.tsx
Comment (regexowl) [inline src/Components/CreateImageWizardV2/steps/Review/Footer/CreateDropdown.tsx:42]: Just to check - are the three new lines in `CreateDropdown` and `EditDropdown` files intentional?

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImagesButton.tsx:49]: This won't work, hooks (useQuery is a hook) can't be called conditionally, see https://redux-toolkit.js.org/rtk-query/usage/conditional-fetching on details.
I think we could use
```jsx
 } = useGetBlueprintQuery(!!selectedBlueprintId ? { id: selectedBlueprintId } : skipToken)
```

**Comment by ezr-ondrej** [comment]
Do we have a mock for this, or will we get ack from UX on the PR screenshot/recording?

**Comment by mgold1234** [comment]
@ezr-ondrej link to figma -https://www.figma.com/design/9t3owYlJi4M23fUXWpgXCE/HMS-Image-Builder?node-id=970-3284&t=8C1COFOncZHlDngQ-0

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImagesButton.tsx:125]: could we use the `imageType` as key and itemId and have selectedItems contain the imageType directly, not an index? I'm just thinking out loud, but might be bit safer (the order of blueprintImageType can not be relied on 100% IMO :) )

**Comment by regexowl** [comment]
Went through the mocks and what the PR looks like right now:
- the mocks seem to work on subtraction, meaning all the check boxes will be checked when the drop down menu is opened and you can uncheck environments to remove the target from the build queue. I feel like that makes sense as clicking on "Build images" directly should have the same functionality as it had before, that is building all targets included in the blueprint
image

- the targets have their full name in the drop down, which should be easy to update as we do have a target mapping object in `Target.tsx` which we could export and use
image
image

- when some of the targets are unchecked/removed from the build the text on the button should change from "Build images" to "Build selected", you can probably just compare the length of original list and length of selected list and render the name based on whether they match or something like that
image

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:125]: All for using imageType as key since the values are unique :+1:

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:24]: This component is used in two places:

image

When the props were changed a conflict between the attributes was introduced, causing this error:
image

Originally the component was called with variant (which should be "primary" and "link" from top to bottom) and children.

There are two ways to resolve this, either we can try to bend the current logic to be reusable with the other "Build images" button or the link "Build images" button can be added independent of the original component since now it wouldn't be a code duplication as the buttons are quite different.

Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:24]: This is how the second button should look like:
image

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:40]: Since you have `useState` in imports you can use just this here:

```suggestion
  const [selectedItems, setSelectedItems] = useState<number[]>([]);
  const [isOpen, setIsOpen] = useState(false);
```

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:61]: We should stay away from type assertion whenever possible as those only allow space for mistakes and future bugs.

The prop types say that `itemId` can be number, string or undefined, but then the const says it's always number. There is no exception handling so if itemId returns as undefined, the service could crash as this line would try to index with undefined:
```
const imageType = blueprintImageType?.[item];
```

But that being said I agree with Ondřej that we could and probably should index with the imageType, making the type string and going from there :)

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:57]: Let's find out more descriptive name - maybe `onSelectTarget`? Or whatever you think will be more fitting :)

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:44]: Or wait - do we even need `onSelect` here? Don't these lines do the same thing?

```
onSelect={onSelect}
onOpenChange={(isOpen: boolean) => setIsOpen(isOpen)}
```

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImagesButton.tsx:25]: IMO this means it's all unselected by default, IMO easier would be to
```suggestion
  const [selectedTargets, setSelectedTargets] =useState<ImageTypes[] | undefined>(undefined);
```

**Review by ezr-ondrej — COMMENTED** [review]
I've left few comments I see we should fix to make the selection work. We have two states for it, I think we should  have just one.

File: src/Components/Blueprints/BuildImageButtonEmpytyState.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImageButtonEmpytyState.tsx:14]: IMO this component could live in the same file (BuildImagesButton), there is no strong need to have it separate.

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImagesButton.tsx:129]: This might be better extracted to a separate function.
Might be ok to just have it in this file for now, but ideal would be some utilities/const file.
```js
export const humanReadableImageTarget = (target) =>
  switch(target) {
    case 'aws':
     return 'Amazon Web Services (AWS)';
  .....
```

I also wonder if we don't already have it, we need to have targets human translated.
EDIT: we do not have anything like it yet as far as I could find, so just having such function here localy makes sense :)

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (ezr-ondrej) [inline src/Components/Blueprints/BuildImagesButton.tsx:35]: I'm not sure here, but where do we handle errors? :thinking:
@regexowl / @amirfefer ?

**Review by ezr-ondrej — COMMENTED** [review]
I've not tested it, but the code looks quite reasonable, I've left few improvement ideas for the code readability and error handling, but apart of that looks quite ready :)

It would be nice to add at least some basic tests for it. Do you plan to do so in this PR?

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:108]:
```suggestion
                      : 'Build images'}
```

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:89]: When no targets are selected the "Build images" option is still available and ends up in a toast notifications "Images are being built", but nothing else happens. Let's disable the button when all the targets are deselected.

```suggestion
                isDisabled={
                  !selectedBlueprintId ||
                  selectedTargets.length === blueprintImageType?.length
                }
```

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:30]: Nit pick, but since the logic works on subtraction would it make sense to call this "deselectedTargets"?

**Review by regexowl — COMMENTED** [review]
This looks really good, thanks for working on this! Added some comments :)

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:35]: That's the neat part - we don't... but in reality we probably should. It would be similar to the loading logic, what needs to be done is adding error to the mutation:

```typescript
  const [buildBlueprint, { isLoading: imageBuildLoading, isError: imageBuildError }] =
    useComposeBlueprintMutation();
```

And then do something like this probably?

```typescript
  try {
    await buildBlueprint({
      ...
  catch (imageBuildError) {
    dispatch(
      addNotification({
        <something's wrong>
```

Seems like we seldom properly handle errors in mutations though. :face_with_peeking_eye:

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:129]: Extracting the replace logic into separate function sounds good. I have an idea on how I'd like to deal with the mapping throughout the code base so I'd open a follow up PR once this is merged.

**Review by ezr-ondrej — COMMENTED** [review]
Nice first test!
Should we also test, that when selecting only one target out of multiple targets that are available, the correct request is send to a server?

File: src/test/fixtures/blueprints.ts
Comment (ezr-ondrej) [inline src/test/fixtures/blueprints.ts:23]: linter complaints.

File: src/Components/Blueprints/BuildImagesButton.tsx
Comment (regexowl) [inline src/Components/Blueprints/BuildImagesButton.tsx:100]: https://github.com/osbuild/image-builder-frontend/pull/2135 got merged today - you can use `targetOptions` const to replace this function so the naming is consistent throughout the code base: https://github.com/osbuild/image-builder-frontend/blob/ee216857926bf1a628e74733bd4bf7942b7ad4e3/src/constants.ts#L51-L67

Comment (regexowl) [reply src/Components/Blueprints/BuildImagesButton.tsx:100]: But we can split it to a follow up as well, whatever works better :) This is just a heads up

File: src/test/Components/Blueprints/Blueprints.test.tsx
Comment (ezr-ondrej) [inline src/test/Components/Blueprints/Blueprints.test.tsx:128]: The test doesn't seem to test what it says it's testing, or does it?
I see expect on button to be disabled, that's probably not correct, right?

**Comment by InsightsDroid** [comment]
## :exclamation: Outdated dependencies error
* Installed version of @redhat-cloud-services/frontend-components-config (5.0.5) is older than required version (6.0.0)

**Comment by InsightsDroid** [comment]
## :exclamation: Outdated dependencies error
* Installed version of @redhat-cloud-services/frontend-components-config (5.0.5) is older than required version (6.0.0)

**Review by regexowl — APPROVED** [review]
Thank you! :heart:

**Comment by InsightsDroid** [comment]
## :exclamation: Outdated dependencies error
* Installed version of @redhat-cloud-services/frontend-components-config (5.0.5) is older than required version (6.0.0)

**Pull request merged**
